overlord/devicestate/devicemgr.go

```diff
@@ -1206,16 +1224,51 @@ func (m *DeviceManager) ensureFactoryReset() error {
 
 	perfTimings := timings.New(map[string]string{"ensure": "factory-reset"})
 
+	model, err := m.Model()
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+	if err != nil {
+		return fmt.Errorf("internal error: core device brand and model are set but there is no model assertion")
+	}
+
+	// We perform this check before setting ensureFactoryResetRan in
+	// case this should fail. This should in theory not be possible as
+	// the same type of check is made during install-mode.
+	hasInstallDeviceHook, err := m.hasInstallDeviceHook(model)
+	if err != nil {
+		return fmt.Errorf("internal error: %v", err)
+	}
+
 	m.ensureFactoryResetRan = true
 
+	// Create both factory-reset-run-system and restart-system-to-run-mode tasks as they
+	// will run unconditionally. They will be chained together with optionally the
+	// install-device hook task.
 	factoryReset := m.state.NewTask("factory-reset-run-system", i18n.G("Perform factory reset of the system"))
 	restartSystem := m.state.NewTask("restart-system-to-run-mode", i18n.G("Ensure next boot to run mode"))
-	restartSystem.WaitFor(factoryReset)
 
-	// TODO: add factory-reset hooks?
+	var prev *state.Task
+	tasks := []*state.Task{factoryReset}
+	addTask := func(t *state.Task) {
+		t.WaitFor(prev)
+		tasks = append(tasks, t)
+		prev = t
+	}
+	prev = factoryReset
```

Comment:
this could be moved to line 1251 if you used `:=`.

Reply:
Fixed :-)